src/main/java/io/confluent/kafkarest/SimpleConsumerPool.java

```diff
@@ -34,37 +35,192 @@
   private final int poolInstanceAvailabilityTimeoutMs;
   private final Time time;
 
-  private final SimpleConsumerFactory simpleConsumerFactory;
-  private final Map<String, SimpleConsumer> simpleConsumers;
-  private final Queue<String> availableConsumers;
+  private final int maxPollTime;
 
-  public SimpleConsumerPool(int maxPoolSize, int poolInstanceAvailabilityTimeoutMs,
-                            Time time, SimpleConsumerFactory simpleConsumerFactory) {
+  private final ConsumerFactory<byte[], byte[]> simpleConsumerFactory;
+  private final Map<TopicPartition, Map<Long, List<Consumer<byte[], byte[]>>>> topicPartitionOffsetConsumers;
+  private int consumerSize;
+
+  public SimpleConsumerPool(int maxPoolSize,
+                            int poolInstanceAvailabilityTimeoutMs,
+                            int maxPollTime,
+                            Time time, ConsumerFactory<byte[], byte[]> simpleConsumerFactory) {
     this.maxPoolSize = maxPoolSize;
     this.poolInstanceAvailabilityTimeoutMs = poolInstanceAvailabilityTimeoutMs;
     this.time = time;
+    this.maxPollTime = maxPollTime;
+
     this.simpleConsumerFactory = simpleConsumerFactory;
 
-    simpleConsumers = new HashMap<String, SimpleConsumer>();
-    availableConsumers = new LinkedList<String>();
+    // using LinkedHashMap guarantees FIFO iteration order
+    topicPartitionOffsetConsumers = new LinkedHashMap<>();
+    consumerSize = 0;
+  }
+
+  public RecordsFetcher getRecordsFetcher(TopicPartition topicPartition) {
+    return new RecordsFetcher(topicPartition);
+  }
+
+  public final class RecordsFetcher implements AutoCloseable {
+
+    private Consumer<byte[], byte[]> consumer;
+    private final SimpleConsumerPool ownerPool;
+    private final long maxPollTime;
+    private boolean initialized;
+    private Time time;
+    private final TopicPartition topicPartition;
+
+    RecordsFetcher(TopicPartition topicPartition) {
+      this.ownerPool = SimpleConsumerPool.this;
```

Comment:
nit: Given that these are always available to non-`static` nested classes, is there really a need for storing them separately as fields?